File: packages/monitorme-server-agent-js/example-server-collect/middleware.js
const { trace, context, propagation } = require("@opentelemetry/api");

// Function to extract attributes from request headers
function extractAttributes(headers) {
  return {
    segmentId: headers["x-segment-id"] || "",
    sessionId: headers["x-session-id"] || "",
    userId: headers["x-user-id"] || "",
    triggerRoute: headers["x-trigger-route"] || "",
    requestData: headers["x-request-data"] || "",
  };
}

// Function to set extracted attributes on the current span
function setSpanAttributes(span, attributes) {
  Object.keys(attributes).forEach((key) => {
    if (attributes[key]) {
      span.setAttribute(key, attributes[key]); // Only set non-empty attributes
    }
  });
}

// Function to update the current baggage with new attributes
function updateBaggage(currentBaggage, attributes) {
  let updatedBaggage = {};

  // Iterate over the extracted attributes and add them to the new baggage
  Object.keys(attributes).forEach((key) => {
    if (attributes[key]) {
      // Only add if the value is not empty
      updatedBaggage[key] = { value: attributes[key] };
    }
  });

  return propagation.createBaggage({
    ...currentBaggage.getAllEntries(), // Merge with current baggage
    ...updatedBaggage, // Add new attributes to the baggage
  });
}

// Middleware to apply baggage and set attributes on the span
function applyBaggageMiddleware(req, res, next) {
  const attributes = extractAttributes(req.headers); // Extract attributes from request headers

  const activeContext = context.active(); // Get the current active context (associated with the request)
  const currentSpan = trace.getSpan(activeContext); // Get the current active span for tracing

  if (currentSpan) {
    // Set the extracted attributes on the current span for local trace visibility
    setSpanAttributes(currentSpan, attributes);
  }

  // Get the current baggage (if any) or create a new baggage if none exists
  const currentBaggage =
    propagation.getBaggage(activeContext) || propagation.createBaggage();

  // Update the current baggage by adding the new attributes to it
  const updatedBaggage = updateBaggage(currentBaggage, attributes);

  // Propagate the updated baggage in the active context, allowing it to be passed
  // to the next service (Service C, etc.)
  context.with(propagation.setBaggage(activeContext, updatedBaggage), next);
}

module.exports = applyBaggageMiddleware;

File: packages/monitorme-server-agent-py/monitorme_server_agent_py/middleware.py
from __future__ import annotations

from collections.abc import Mapping

from opentelemetry import baggage, context, trace

HEADER_TO_ATTRIBUTE = {
    "x-segment-id": "segmentId",
    "x-session-id": "sessionId",
    "x-user-id": "userId",
    "x-trigger-route": "triggerRoute",
    "x-request-data": "requestData",
}


def _extract_attributes(headers: Mapping[str, str]) -> dict[str, str]:
    attributes: dict[str, str] = {}
    for header_name, attribute_name in HEADER_TO_ATTRIBUTE.items():
        value = headers.get(header_name, "").strip()
        if value:
            attributes[attribute_name] = value
    return attributes


def _set_span_attributes(attributes: dict[str, str]) -> None:
    current_span = trace.get_current_span()
    if current_span and current_span.is_recording():
        for key, value in attributes.items():
            current_span.set_attribute(key, value)


def _attach_baggage(attributes: dict[str, str]) -> context.Token:
    current_context = context.get_current()
    for key, value in attributes.items():
        current_context = baggage.set_baggage(key, value, context=current_context)
    return context.attach(current_context)


class MonitorMeASGIBaggageMiddleware:
    def __init__(self, app) -> None:
        self.app = app

    async def __call__(self, scope, receive, send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        headers = {
            key.decode("latin-1").lower(): value.decode("latin-1")
            for key, value in scope.get("headers", [])
        }
        attributes = _extract_attributes(headers)

        if not attributes:
            await self.app(scope, receive, send)
            return

        _set_span_attributes(attributes)
        token = _attach_baggage(attributes)
        try:
            await self.app(scope, receive, send)
        finally:
            context.detach(token)
